### ISSUE TITLE
Unable to make concurrent requests

### ISSUE THREAD

**Issue by natdm** [stripe/stripe-go#469]
Latest version of stripe-go

When I make concurrent requests to get orders, I get a panic due to `getCachedOrBuildStructEncoder` reading from the `structCache` at the same time as it may be writing to it. The same thing happens in `getCachedOrBuildTypeEncoder` ~~~and maybe a few other places~~~.

When I fix this manually by wrapping the cache from this:
```go
func getCachedOrBuildStructEncoder(t reflect.Type) *structEncoder {
	m, _ := structCache.value.Load().(map[reflect.Type]*structEncoder)
	f := m[t]
	if f != nil {
		return f
	}

	// We do the work to get the encoder without holding a lock. This could
	// result in duplicate work, but it will help us avoid a deadlock. Encoders
	// may be built and stored recursively in the cases of something like an
	// array or slice, so we need to make sure that this function is properly
	// re-entrant.
	f = makeStructEncoder(t)

	structCache.mu.Lock()
	defer structCache.mu.Unlock()

	m, _ = structCache.value.Load().(map[reflect.Type]*structEncoder)
	newM := m
	if newM == nil {
		newM = make(map[reflect.Type]*structEncoder)
		structCache.value.Store(newM)
	}
	newM[t] = f

	return f
}
```

to this:
```go
func getCachedOrBuildStructEncoder(t reflect.Type) *structEncoder {
	structCache.mu.Lock()
	m, _ := structCache.value.Load().(map[reflect.Type]*structEncoder)
	structCache.mu.Unlock()
	f := m[t]
	if f != nil {
		return f
	}

	// We do the work to get the encoder without holding a lock. This could
	// result in duplicate work, but it will help us avoid a deadlock. Encoders
	// may be built and stored recursively in the cases of something like an
	// array or slice, so we need to make sure that this function is properly
	// re-entrant.
	f = makeStructEncoder(t)

	structCache.mu.Lock()
	defer structCache.mu.Unlock()

	m, _ = structCache.value.Load().(map[reflect.Type]*structEncoder)
	newM := m
	if newM == nil {
		newM = make(map[reflect.Type]*structEncoder)
		structCache.value.Store(newM)
	}
	newM[t] = f

	return f
}
```
there is no more panic.

Here is the bit of the stacktrace that should be able to help:
```
fatal error: concurrent map read and map write

goroutine 96 [running]:
runtime.throw(0x17a2578, 0x21)
	/usr/local/go/src/runtime/panic.go:605 +0x95 fp=0xc42cd016c0 sp=0xc42cd016a0 pc=0x102c035
runtime.mapaccess1(0x16b8ca0, 0xc4201a5230, 0xc42cd01770, 0x1042dc4)
	/usr/local/go/src/runtime/hashmap.go:355 +0x238 fp=0xc42cd01708 sp=0xc42cd016c0 pc=0x100a168
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.getCachedOrBuildTypeEncoder(0x1b696a0, 0x1673be0, 0x0)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:255 +0xc2 fp=0xc42cd01790 sp=0xc42cd01708 pc=0x137c612
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.makeStructEncoder(0x1b696a0, 0x1742c80, 0xc42ce1b9d8)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:383 +0x307 fp=0xc42cd01970 sp=0xc42cd01790 pc=0x137d7b7
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.getCachedOrBuildStructEncoder(0x1b696a0, 0x1742c80, 0x0)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:234 +0xf1 fp=0xc42cd019f8 sp=0xc42cd01970 pc=0x137c351
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.buildStructEncoder(0x1b696a0, 0x1742c80, 0xc42049c480)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:204 +0x39 fp=0xc42cd01a30 sp=0xc42cd019f8 pc=0x137bd39
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.makeTypeEncoder(0x1b696a0, 0x1742c80, 0xc42ce1bac0)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:419 +0x1cc fp=0xc42cd01a58 sp=0xc42cd01a30 pc=0x137deac
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.getCachedOrBuildTypeEncoder(0x1b696a0, 0x1742c80, 0x0)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:265 +0xf1 fp=0xc42cd01ae0 sp=0xc42cd01a58 pc=0x137c641
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.buildPtrEncoder(0x1b696a0, 0x16f72e0, 0x1000)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:193 +0x4d fp=0xc42cd01b18 sp=0xc42cd01ae0 pc=0x137bc7d
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.makeTypeEncoder(0x1b696a0, 0x16f72e0, 0xc42ce1bba8)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:413 +0x15a fp=0xc42cd01b40 sp=0xc42cd01b18 pc=0x137de3a
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.getCachedOrBuildTypeEncoder(0x1b696a0, 0x16f72e0, 0x0)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:265 +0xf1 fp=0xc42cd01bc8 sp=0xc42cd01b40 pc=0x137c641
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.reflectValue(0xc42cd80e00, 0x16f72e0, 0xc42cd7a0c0, 0x16, 0x0, 0x0, 0x0)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:326 +0x65 fp=0xc42cd01c28 sp=0xc42cd01bc8 pc=0x137d375
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form.AppendTo(0xc42cd80e00, 0x16f72e0, 0xc42cd7a0c0)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/form/form.go:113 +0x7c fp=0xc42cd01c70 sp=0xc42cd01c28 pc=0x137b31c
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/order.Client.Get(0x1b616a0, 0xc42cd0d170, 0xc4202575a0, 0x20, 0xc42d33e200, 0x1b, 0xc42cd7a0c0, 0x0, 0xc420023900, 0xc42cc6f580)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/order/client.go:107 +0x1b1 fp=0xc42cd01d08 sp=0xc42cd01c70 pc=0x142e091
github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/order.Get(0xc42d33e200, 0x1b, 0xc42cd7a0c0, 0x15e43b7, 0xc42cc6f5d8, 0xc42cc6f5e0)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/vendor/github.com/stripe/stripe-go/order/client.go:97 +0xaf fp=0xc42cd01d90 sp=0xc42cd01d08 pc=0x142ddcf
github.com/natdm/bideasy/billing.(*Billing).GetOrder(0xc420192b00, 0xc42d13dd80, 0x15, 0xc42d33e200, 0x1b, 0x0, 0x0, 0x0)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/billing/billing.go:183 +0x73 fp=0xc42cd01dd8 sp=0xc42cd01d90 pc=0x1433d03
github.com/natdm/bideasy/controllers.(*Admin).Orders.func2(0xc42d234e40, 0xc4200f6180, 0xc42d0b89a0, 0xc42d19cd00, 0xc42006d500, 0x4, 0x2415cee8, 0xed162ff87, 0x0, 0x2415cee8, ...)
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/controllers/admin.go:1621 +0xb5 fp=0xc42cd01f40 sp=0xc42cd01dd8 pc=0x15e4415
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2337 +0x1 fp=0xc42cd01f48 sp=0xc42cd01f40 pc=0x10592c1
created by github.com/natdm/bideasy/controllers.(*Admin).Orders
	/Users/nathanhyland/gocode/src/github.com/natdm/bideasy/controllers/admin.go:1619 +0x496
```

I was going to make a PR but I didn't know if someone had a different solution than just wrapping it in the locks.

**Comment by brandur-stripe** [comment]
Doh, I should've written a concurrency test. Thanks for logging! I'll take a look tomorrow.

**Comment by brandur-stripe** [comment]
Hi @natdm, thanks again for the detail on your issue here.

I'm introducing a fix in #470. It's similar to the one you proposed except that I use an `RWMutex` instead, which is a slightly nicety because it means that after an initial value is cached, unlimited readers are able to access values on that map simultaneously, so there should be very little overhead to concurrent operations. I also stripped out all the code around `atomic.Value` now that all operations on the cache map are wrapped in a mutex.

**Comment by natdm** [comment]
I appreciate the speed of response/delivery you guys have on issues. Bot you and your support line.
